src/netaddress.h

```diff
@@ -220,13 +212,21 @@ class CNetAddr
         return IsIPv4() || IsIPv6() || IsTor() || IsI2P() || IsCJDNS();
     }
 
+    enum class Encoding {
+        V1,
+        V2, //!< BIP155 encoding
+    };
+    struct SerParams {
+        const Encoding enc;
+    };
+
     /**
      * Serialize to a stream.
      */
     template <typename Stream>
     void Serialize(Stream& s) const
     {
-        if (s.GetVersion() & ADDRV2_FORMAT) {
+        if (s.GetParams().enc == Encoding::V2) {
```

Comment:
If the stream has parameters but not the right parameters this gives a mildly confusing error. Writing it as:

```c++
    template <typename Stream>
    void Serialize(Stream& s, const SerParams& params) const

    template <typename Stream> void Serialize(Stream& s) const { return Serialize(s, s.GetParams()); }
```

seems a bit nicer. It seems like there should be some way to have `::Serialize()` call this with the parameters directly via some template magic, but I can't see what it would be.

Reply:
Yeah; with C++20 concepts I could get `::Serialize` to call `obj.Serialize(s, s.GetParams())` directly which seemed nicer, but with just C++17 I got myself too confused.